File: examples/src/test/scala/docs/testkit/TestKitDocExample.scala
/*
 * Copyright (C) 2020-2021 Lightbend Inc. <https://www.lightbend.com>
 */

package docs.testkit

import scala.concurrent.Future

import akka.Done
import akka.projection.ProjectionId
import akka.projection.scaladsl.Handler
import akka.projection.testkit.scaladsl.TestProjection
import akka.projection.testkit.scaladsl.TestSourceProvider

//#testkit-import
import akka.actor.testkit.typed.scaladsl.ScalaTestWithActorTestKit
import akka.projection.testkit.scaladsl.ProjectionTestKit

//#testkit-import

//#testkit-duration
import scala.concurrent.duration._

//#testkit-duration

//#testkit-testprojection
import akka.stream.scaladsl.Source

//#testkit-testprojection

// FIXME: this test can't run, most of its building blocks are 'fake' or are using null values
// abstract so it doesn run for now
abstract
//#testkit
class TestKitDocExample extends ScalaTestWithActorTestKit {
  private val projectionTestKit = ProjectionTestKit(system)

  //#testkit

  case class CartView(id: String)

  class CartViewRepository {
    def findById(id: String): Future[CartView] = Future.successful(CartView(id))
  }

  val cartViewRepository = new CartViewRepository

  // it only needs to compile
  val projection = TestProjection(ProjectionId("test", "00"), sourceProvider = null, handler = null)

  {
    //#testkit-run
    projectionTestKit.run(projection) {
      // confirm that cart checkout was inserted in db
      cartViewRepository.findById("abc-def").futureValue
    }
    //#testkit-run
  }

  {
    //#testkit-run-max-interval
    projectionTestKit.run(projection, max = 5.seconds, interval = 300.millis) {
      // confirm that cart checkout was inserted in db
      cartViewRepository.findById("abc-def").futureValue
    }
    //#testkit-run-max-interval
  }

  //#testkit-sink-probe
  projectionTestKit.runWithTestSink(projection) { sinkProbe =>
    sinkProbe.request(1)
    sinkProbe.expectNext(Done)
  }

  // confirm that cart checkout was inserted in db
  cartViewRepository.findById("abc-def").futureValue

  //#testkit-sink-probe

  {
    val handler: Handler[(Int, String)] = null

    //#testkit-testprojection
    val testData = Source((0, "abc") :: (1, "def") :: Nil)

    val extractOffset = (envelope: (Int, String)) => envelope._1

    val sourceProvider = TestSourceProvider(testData, extractOffset)

    val projection = TestProjection(ProjectionId("test", "00"), sourceProvider, () => handler)

    projectionTestKit.run(projection) {
      // assert logic ..
    }
    //#testkit-testprojection
  }
//#testkit
}
//#testkit

File: examples/src/test/java/jdocs/testkit/TestKitDocExample.java
/*
 * Copyright (C) 2020-2025 Lightbend Inc. <https://www.lightbend.com>
 */

package jdocs.testkit;

import akka.Done;
import akka.NotUsed;
import akka.projection.Projection;
import akka.projection.ProjectionId;
import akka.projection.javadsl.Handler;

import java.util.List;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionStage;

// #testkit-import
import akka.projection.testkit.javadsl.TestSourceProvider;
import org.junit.ClassRule;
import akka.actor.testkit.typed.javadsl.TestKitJunitResource;
import akka.projection.testkit.javadsl.ProjectionTestKit;

// #testkit-import

// #testkit-duration
import java.time.Duration;
import java.util.concurrent.TimeUnit;
import java.util.function.Function;
import java.util.stream.Collectors;
import java.util.stream.Stream;

// #testkit-duration

// #testkit-assertion-import
import static org.junit.Assert.assertEquals;

// #testkit-assertion-import

// #testkit-testprojection

import akka.japi.Pair;
import akka.stream.javadsl.Source;
import akka.projection.testkit.javadsl.TestProjection;

// #testkit-testprojection

public class TestKitDocExample {

  static class CartView {
    final String id;

    CartView(String id) {
      this.id = id;
    }
  }

  static class CartCheckoutRepository {
    public CompletionStage<CartView> findById(String id) {
      return CompletableFuture.completedFuture(new CartView(id));
    }
  }

  // #testkit
  @ClassRule static final TestKitJunitResource testKit = new TestKitJunitResource();
  ProjectionTestKit projectionTestKit = ProjectionTestKit.create(testKit.system());
  // #testkit

  Projection<String> projection = TestProjection.create(null, null, null);

  CartCheckoutRepository cartCheckoutRepository = new CartCheckoutRepository();

  void illustrateTestKitRun() {
    // #testkit-run
    projectionTestKit.run(
        projection,
        () ->
            cartCheckoutRepository
                .findById("abc-def")
                .toCompletableFuture()
                .get(1, TimeUnit.SECONDS));
    // #testkit-run
  }

  void illustrateTestKitRunWithMaxAndInterval() {
    // #testkit-run-max-interval
    projectionTestKit.run(
        projection,
        Duration.ofSeconds(5),
        Duration.ofMillis(300),
        () ->
            cartCheckoutRepository
                .findById("abc-def")
                .toCompletableFuture()
                .get(1, TimeUnit.SECONDS));
    // #testkit-run-max-interval
  }

  void illustrateTestKitRunWithTestSink() {

    // #testkit-sink-probe
    projectionTestKit.runWithTestSink(
        projection,
        sinkProbe -> {
          sinkProbe.request(1);
          sinkProbe.expectNext(Done.getInstance());
          cartCheckoutRepository.findById("abc-def").toCompletableFuture().get(1, TimeUnit.SECONDS);
        });

    // #testkit-sink-probe
  }

  void illustrateTestKitTestProjection() {
    Handler<Pair<Integer, String>> handler = null;

    // #testkit-testprojection
    List<Pair<Integer, String>> testData =
        Stream.of(Pair.create(0, "abc"), Pair.create(1, "def")).collect(Collectors.toList());

    Source<Pair<Integer, String>, NotUsed> source = Source.from(testData);

    Function<Pair<Integer, String>, Integer> extractOffsetFn =
        (Pair<Integer, String> env) -> env.first();

    TestSourceProvider<Integer, Pair<Integer, String>> sourceProvider =
        TestSourceProvider.create(source, extractOffsetFn);

    Projection<Pair<Integer, String>> projection =
        TestProjection.create(ProjectionId.of("test", "00"), sourceProvider, () -> handler);

    projectionTestKit.run(
        projection,
        () -> {
          // assert logic ...
        });
    // #testkit-testprojection
  }

  // #fixme
  // FIXME: Java example
  // #fixme
}
